### PR TITLE
1019/1017-grouping_days_since_covid

### PR DESCRIPTION
**Pull request by ohibo**
## Pre-review checklist

I have ensured that:

- [x] Code runs successfully in the development environment.
- [x] New code is tested to prove that it works. If not, justify reasoning.
- [X] New code is documented using [type hinting](https://realpython.com/lessons/type-hinting/) and following the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) docstring format.
- [x] New code follows the project naming standards.

## Post-review checklist

As I merge these changes, I have:

- [x] Updated the related Jira ticket, including approximate time spent (including review).
- [x] Moved the related Jira ticket to Done.

### CODE REVIEW

File: cishouseholds/derive.py [hunk 1]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
         + F.round((F.col("month_more") + F.col("day_more")) / 3, 0).cast("int"),
     )
     return df.drop("month_more", "day_more")
+
+
+def assign_days_since_think_had_covid_group(
+    df: DataFrame,
+    think_had_covid_column: str = "think_had_covid",
+    reference_column: str = "days_since_think_had_covid",
+    column_name_to_assign: str = "days_since_think_had_covid_group",
```
Comment (osborj) [inline cishouseholds/derive.py:881]: days_since_reference_column: str (instead of reference_column)
binary_reference_column: str (instead of think_had_covid_column, as also uses contact_any_covid)

Need both, as this function is for days_since_think_had_covid_group and contact_known_or_suspected_covid_days_since_group

File: cishouseholds/derive.py [hunk 2]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
         + F.round((F.col("month_more") + F.col("day_more")) / 3, 0).cast("int"),
     )
     return df.drop("month_more", "day_more")
+
+
+def assign_days_since_think_had_covid_group(
```
Comment (osborj) [inline cishouseholds/derive.py:877]: Change function name as applicable to more than one variable

assign_days_since_grouped_variable

File: cishouseholds/derive.py [hunk 3]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
...
+    contact_known_or_suspected_covid_days_since_group. The variable
+    days_since_think_had_covid will give a number that will be grouped
+    in a range so long think_had_covid_column is positive, otherwise will
+    be None.
+    Parameters
+    ----------
+    df
+    think_had_covid_column
+        yes/no values that describe whether the patient thinks have had covid
+    reference_column
```
Comment (osborj) [inline cishouseholds/derive.py:894]: Change these definitions

File: cishouseholds/derive.py [hunk 4]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
...
+def assign_days_since_think_had_covid_group(
+    df: DataFrame,
+    think_had_covid_column: str = "think_had_covid",
+    reference_column: str = "days_since_think_had_covid",
+    column_name_to_assign: str = "days_since_think_had_covid_group",
+) -> DataFrame:
+    """
+    Function to be applied for days_since_think_had_covid_group and
+    contact_known_or_suspected_covid_days_since_group. The variable
+    days_since_think_had_covid will give a number that will be grouped
```
Comment (osborj) [inline cishouseholds/derive.py:886]: This description needs to be generic if applying to derive both variables. Currently referring to deriving one of the two variables

File: cishouseholds/derive.py [hunk 5]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
...
+    """
+    df = assign_named_buckets(
+        df=df,
+        reference_column=reference_column,
+        column_name_to_assign=column_name_to_assign,
+        map={0: "0-14", 15: "15-28", 29: "29-60", 61: "61-90", 91: "91+", None: "Date not given"},
+    )
+    return df.withColumn(
+        column_name_to_assign,
+        F.when(F.col(think_had_covid_column) == "Yes", F.col(column_name_to_assign)).otherwise(None).cast("string"),
```
Comment (osborj) [inline cishouseholds/derive.py:908]:
```suggestion
        F.when((F.col(think_had_covid_column) == "Yes") & (F.col(reference_column).isNull()), "Date not given").otherwise(F.col(column_name_to_assign)).cast("string"),
```

File: cishouseholds/derive.py [hunk 6]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
...
+        column from which extract the number of days transcurred that needs to
+        be grouped
+    column_name_to_assign
+        grouping column
+    """
+    df = assign_named_buckets(
+        df=df,
+        reference_column=reference_column,
+        column_name_to_assign=column_name_to_assign,
+        map={0: "0-14", 15: "15-28", 29: "29-60", 61: "61-90", 91: "91+", None: "Date not given"},
```
Comment (osborj) [inline cishouseholds/derive.py:904]: Remove the None map

File: tests/derive/test_assign_days_since_think_had_covid_group.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+from chispa import assert_df_equality
+
+from cishouseholds.derive import assign_days_since_think_had_covid_group
+
+
+def test_assign_days_since_think_had_covid_group(spark_session):
+    expected_df = spark_session.createDataFrame(
+        data=[
+            ("yes", 7, "0-14"),
```
Comment (osborj) [inline tests/derive/test_assign_days_since_think_had_covid_group.py:9]: Capitalise these

File: tests/derive/test_assign_days_since_think_had_covid_group.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            ("yes", 20, "15-28"),
+            ("yes", 30, "29-60"),
+            ("yes", 70, "61-90"),
+            ("yes", 100, "91+"),
+            ("Yes", None, "Date not given"),
+            ("no", 6, None),
```
Comment (osborj) [inline tests/derive/test_assign_days_since_think_had_covid_group.py:15]: 'Don't care' on 'think_had_covid' columns if days_since column has a value, so these wouldn't derive None.

File: tests/derive/test_assign_days_since_think_had_covid_group.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+def test_assign_days_since_think_had_covid_group(spark_session):
+    expected_df = spark_session.createDataFrame(
+        data=[
+            ("yes", 7, "0-14"),
+            ("yes", 20, "15-28"),
+            ("yes", 30, "29-60"),
+            ("yes", 70, "61-90"),
+            ("yes", 100, "91+"),
+            ("Yes", None, "Date not given"),
```
Comment (osborj) [inline tests/derive/test_assign_days_since_think_had_covid_group.py:14]: Add test here for ("No", None, None)

File: cishouseholds/derive.py [hunk 10]
```diff
@@ -872,3 +872,38 @@ def assign_correct_age_at_date(df: DataFrame, column_name_to_assign, reference_d
         + F.round((F.col("month_more") + F.col("day_more")) / 3, 0).cast("int"),
     )
     return df.drop("month_more", "day_more")
+
```
Comment (osborj) [inline cishouseholds/derive.py:875]: Is it also needed to include the derivation of these into the ETL scripts?
Currently not deriving the two variables.

**Review by osborj — CHANGES_REQUESTED** [review]
